ghcide/src/Development/IDE/Core/Rules.hs

```diff
@@ -650,7 +650,11 @@ typeCheckRule :: Recorder (WithPriority Log) -> Rules ()
 typeCheckRule recorder = define (cmapWithPrio LogShake recorder) $ \TypeCheck file -> do
     pm <- use_ GetParsedModule file
     hsc  <- hscEnv <$> use_ GhcSessionDeps file
-    typeCheckRuleDefinition hsc pm
+    foi <- use_ IsFileOfInterest file
+    -- We should only call the typecheck rule for files of interest.
+    -- Keeping typechecked modules in memory for other files is
+    -- very expensive.
+    assert (foi /= NotFOI) $ typeCheckRuleDefinition hsc pm
```

Comment:
I have this assertion locally and it often fails for me but not necessarily in bad ways, just seems there is a race?

Reply:
Can you get a call stack?

Reply:
If you close a file while a typecheck has been queued, then on restart that typecheck will be executed, but the assertion will fail because it is not an FOI any longer.

I think this is harmless because it should be garbage collected soon after?

Reply:
Sounds like a good reason to make it just a warning :)

Reply:
Yes, but I think the false positive rate can be very high to the point of being annoying. Maybe we should just silently log it, with a message that explains this caveat and asks for a bug report. 

Something to effect of "Warning: Typecheck executed for file that is not currently open. This is harmless if this file was recently open and has now been closed. If not, please report a bug."

Reply:
> If you close a file while a typecheck has been queued,

Can the check be moved to the code that schedules the typecheck?